File: src-tauri/src/events.rs
use serde::Serialize;

// Event name constants
pub const EVENT_RECORDING_STATE_CHANGED: &str = "recording-state-changed";
pub const EVENT_TRANSCRIPTION_COMPLETE: &str = "transcription-complete";
pub const EVENT_DOWNLOAD_PROGRESS: &str = "download-progress";
pub const EVENT_ERROR: &str = "app-error";

// Event payload types
#[derive(Debug, Clone, Serialize)]
pub struct DownloadProgress {
    pub model_id: String,
    pub downloaded_bytes: u64,
    pub total_bytes: u64,
    pub percent: f32,
}

File: src/lib/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type { DownloadProgress, HistoryEntry, RecordingState } from "../types";

export const onRecordingStateChanged = (
  cb: (state: RecordingState) => void,
): Promise<UnlistenFn> =>
  listen<RecordingState>("recording-state-changed", (e) => cb(e.payload));

export const onTranscriptionComplete = (
  cb: (entry: HistoryEntry) => void,
): Promise<UnlistenFn> =>
  listen<HistoryEntry>("transcription-complete", (e) => cb(e.payload));

export const onDownloadProgress = (
  cb: (progress: DownloadProgress) => void,
): Promise<UnlistenFn> =>
  listen<DownloadProgress>("download-progress", (e) => cb(e.payload));

export const onSettingsUpdated = (cb: () => void): Promise<UnlistenFn> =>
  listen("settings-updated", () => cb());

export const onError = (cb: (message: string) => void): Promise<UnlistenFn> =>
  listen<string>("app-error", (e) => cb(e.payload));
